File: 73.set-matrix-zeroes.py
#
# @lc app=leetcode id=73 lang=python
#
# [73] Set Matrix Zeroes
#

# @lc code=start
class Solution(object):
   def setZeroes(matrix):
    # find the rows and columns that contain zeros
    rows = set()
    columns = set()
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if matrix[i][j] == 0:
                rows.add(i)
                columns.add(j)
    
    # set the entire row and column to zeros
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if i in rows or j in columns:
                matrix[i][j] = 0
    
    # return the modified matrix
    return matrix

# test the function with the given input
matrix = [[1,1,1],[1,0,1],[1,1,1]]
print("Input:", matrix)
print("Output:", setZeroes(matrix))

        
# @lc code=end

File: 73.set-matrix-zeroes.cpp
/*
 * @lc app=leetcode id=73 lang=cpp
 *
 * [73] Set Matrix Zeroes
 */

// @lc code=start
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size(); // no of row
        int n = matrix[0].size(); // no of col
        
        int colmatrix = 1;
 for (int i = 0; i<m; i++){
            for(int j = 0; j<n ; j++){
                if (matrix[i][j] == 0){
               
               matrix[i][0] = 0;
               if(j!= 0){
                matrix[0][j] = 0;
               }
                else{
                colmatrix = 0;
            }
            }
        }
 }
     for (int i = 1; i<m; i++){
            for(int j = 1; j<n ; j++){
                if (matrix[i][j] != 0) {
                    if ((matrix[0][j]==0) || (matrix[i][0]==0)) {
                     matrix[i][j] = 0;
                    
                }
      
            }
            }
        } 

       
    if (matrix[0][0] == 0){
        for (int i = 0; i<n; i++){
            matrix[0][i] = 0;
            }

    }
     if (colmatrix == 0){
        for (int i = 0; i<m; i++){
            matrix[i][0] = 0;

        }
    }
  
  
    }
};
// @lc code=end

